### ISSUE TITLE
Arbitrage bot

### ISSUE THREAD

**Issue by rndquu** [ubiquity/ubiquity-dollar#958]
In order to maintain [Dollar token](https://github.com/ubiquity/ubiquity-dollar/blob/0a230f977ad91bbfb01dd28bc1d2ea77faed5237/packages/contracts/src/dollar/core/UbiquityDollarToken.sol) USD peg [LibUbiquityPool](https://github.com/ubiquity/ubiquity-dollar/blob/0a230f977ad91bbfb01dd28bc1d2ea77faed5237/packages/contracts/src/dollar/libraries/LibUbiquityPool.sol) provides the following arbitrage mechanic:
1. If Dollar token USD price >1$ (on a secondary market) then arbitragers are incentivised to 
a) Mint 1 Dollar token for 1 USD of collateral in the [LibUbiquityPool](https://github.com/ubiquity/ubiquity-dollar/blob/0a230f977ad91bbfb01dd28bc1d2ea77faed5237/packages/contracts/src/dollar/libraries/LibUbiquityPool.sol)
b) Sell 1 Dollar token on a secondary market
2. On the contrary, If Dollar token USD price <1$ (on a secondary market) then arbitragers are incentivised to:
a) Buy cheap Dollar token on a secondary market
b) Redeem Dollar token for collateral worth 1 USD in the [LibUbiquityPool](https://github.com/ubiquity/ubiquity-dollar/blob/0a230f977ad91bbfb01dd28bc1d2ea77faed5237/packages/contracts/src/dollar/libraries/LibUbiquityPool.sol)

On early project stage any large swap may affect the Dollar token price greatly and the pool (where the swap occurred) may remain unbalanced for a long time which in turn depegs the Dollar token.

As a part of this issue we should create an arbitrage bot (i.e. `nodejs` app) which could be run on VPS or any suitable PAAS (heroku, cloudflare, etc...) to catch arbitrage opportunities in order to keep the Dollar token USD pegged on early project stage.

What should be done:
1. Create a new `nodejs` app that will catch arbitrage opportunities for the Dollar token
2. Add unit tests for basic functionality
3. Prepare a clear README of how to run the bot

Notes:
- It makes sense to create the arbitrage bot in a separate repository.
- Deployed contract addresses (mainnet): [here](https://github.com/ubiquity/ubiquity-dollar/blob/0a230f977ad91bbfb01dd28bc1d2ea77faed5237/packages/contracts/broadcast/Deploy001_Diamond_Dollar_Governance.s.sol/1/run-latest.json)

**Comment by 0x4007** [comment]
Perhaps we can fork a bot dropping the time estimate down to a day?

**Comment by rndquu** [comment]
> It could be interesting to have a worker spawn on transaction events to the market. Webhook style. But maybe too complicated due to nodejs limitations

Exactly. We need a simple "fork and run" solution because current version of the arbitrage bot will viable only until more sophisticated MEV bots add Dollar token pools to their platforms.

> Perhaps we can fork a bot dropping the time estimate down to a day?

If there already exists a solution, then yes, time estimate can be reduced.

**Comment by rndquu** [comment]
@FibrinLab Did you check the expected flow [here](https://github.com/ubiquity/ubiquity-dollar/issues/958#issuecomment-2345368648) (the bot should be able to run as a github action)?

**Comment by FibrinLab** [comment]
> @FibrinLab Did you check the expected flow [here](https://github.com/ubiquity/ubiquity-dollar/issues/958#issuecomment-2345368648) (the bot should be able to run as a github action)?

Yes. Working on this

**Comment by FibrinLab** [comment]
@rndquu I understand this would need a new repo. Where do I make a draft pull request?

**Comment by rndquu** [comment]
> @rndquu I understand this would need a new repo. Where do I make a draft pull request?

https://github.com/ubiquity/arbitrage-bot

**Comment by 0x4007** [comment]
> > @rndquu I understand this would need a new repo. Where do I make a draft pull request?
> 
> 
> 
> https://github.com/ubiquity/arbitrage-bot

In the future you should make a new repo using our ts-template and host on your own org. Next you work on a pull to your own org but link this task using a closing keyword. 

I don't want to make new repositories until work is completed because I had to delete eight "ts-templates" across our organization due to all the non completes.

**Comment by FibrinLab** [comment]
@rndquu Some updates, created draft pool with arbitrage contract modified to work with Ubq pool. Working on Cron job.

**Comment by surafeldev** [comment]
https://github.com/surafeldev/arbitrage_bot

I made this implementation for this task, let me know if there is any thing I have to consider

**Comment by 0x4007** [comment]
@surafeldev You need to do a pull to your own default branch and link this issue so all of us including the bot can track progress

**Comment by surafeldev** [comment]
as it is stated in the description I have created a separate Repo for the Bot, Do I need to create a PA from bot repo or this repo ?

**Comment by 0x4007** [comment]
Yes you link this issue

**Comment by surafeldev** [comment]
How can I do that ?

**Comment by surafeldev** [comment]
https://github.com/ubiquity/ubiquity-dollar/pull/976
take a look at the repo inside the PA

**Comment by 0x4007** [comment]
You're not using our template and you linked this wrong. I'm not optimistic about the success of this deliverable.

**Comment by surafeldev** [comment]
where can I find the template ? can you share with me

**Comment by surafeldev** [comment]
I couldn`t find the dollar token in Dex to implement the price feature; where can I find the secondary market API? or should I have to build it by expecting it will be implemented ?
